proposals/p2022.md

```diff
@@ -0,0 +1,288 @@
+as unused, but later used.
+
+## Background
+
+See
+[the overall design](https://github.com/carbon-language/carbon-lang/blob/09e4417431b77c05ceb2a98dd38833276514a1ff/docs/design/README.md#binding-patterns)
+for the current state of things with regards to name bindings and the underscore
+character, `_`.
+
+See
+[issue #476: Optional argument names (unused arguments)](https://github.com/carbon-language/carbon-lang/issues/476)
+for the previous conversation on this topic.
+
+Note: These are not exhaustive lists and not intended to be thoroughly
+investigated, rather a brief overview of some prior art.
+
+C and C++ style guides, linters, and other tools have addressed unused function
+parameters specifically with varying levels of clarity for readers.
+
+-   [Standard C++ Foundation's guidelines on unused parameters](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-unused)
+-   [Google's C++ style guide recommendations on unused parameters](https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions)
+-   [C++ `maybe_unused` attribute](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused)
+-   [GCC C `unused` attribute](https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Function-Attributes.html#index-g_t_0040code_007bunused_007d-attribute_002e-2640)
+
+More generally, various languages use the underscore character (`_`) to signal
+an unused binding or pattern.
+
+-   [Rust `_` Patterns](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#ignoring-values-in-a-pattern)
+-   [Golang Blank Identifiers](https://go.dev/ref/spec#Blank_identifier)
+-   [PyLint allows parameters starting with `_`, `ignored_`, or `unused_` to be unused](https://pylint.pycqa.org/en/latest/user_guide/configuration/all-options.html#ignored-argument-names)
+-   [Scala Wildcard Patterns](https://www.scala-lang.org/files/archive/spec/2.11/10-pattern-matching.html#variable-patterns)
+-   [Crystal `_` Cases](https://crystal-lang.org/reference/1.5/syntax_and_semantics/case.html#underscore-allowed)
+
+## Proposal
+
+Users can explicitly declare a value in a pattern as unused in two forms:
+
+-   `_: i32` : Using the underscore token, `_`, in place of a name.
+
+-   `unused size: i32` : Using the leading `unused` keyword followed by the
+    name.
+
+## Details
+
+Introducing two syntaxes satisfies the desire to have a terse way to discard
+values, but still provides authors with a more verbose, explicit syntax that
+preserves the name.
+
+Both approaches are unambiguous -- to human readers and authors as well as
+programmatic interpretations. The inclusion of an explicit `unused` keyword
+allows authors to preserve the name of a value for documentation purposes, while
+still explicitly marking the value as discarded in an interpretable way to
+humans and programs alike.
+
+The `unused` keyword would be a new keyword in Carbon. This keyword would only
```

Comment:
I would like to see a mention somewhere (perhaps "Alternatives considered") of using an attribute rather than a keyword -- whether it's saying "we've considered this and a keyword is the better choice" or to say "we don't have a design for attributes yet and we should consider whether we want to make this an attribute when we do" or even "once we have attributes this should definitely be one of those".